File: src/org/lineageos/xiaomi_bluetooth/settings/configs/NoiseCancellationListController.java
package org.lineageos.xiaomi_bluetooth.settings.configs;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.preference.Preference;

import org.lineageos.xiaomi_bluetooth.EarbudsConstants;
import org.lineageos.xiaomi_bluetooth.R;
import org.lineageos.xiaomi_bluetooth.mma.MMADevice;

import java.io.IOException;
import java.util.List;
import java.util.Set;


public class NoiseCancellationListController extends MultiListController
        implements Preference.OnPreferenceChangeListener {

    private static final String TAG = ListController.class.getSimpleName();
    private static final boolean DEBUG = true;

    private static final byte VALUE_NOT_MODIFIED = -1;

    private static final List<ConfigState> CONFIG_STATES = List.of(
            new ConfigState(new byte[]{0x00}, R.string.noise_cancellation_mode_off),
            new ConfigState(new byte[]{0x01}, R.string.noise_cancellation_mode_on),
            new ConfigState(new byte[]{0x02}, R.string.noise_cancellation_mode_transparency)
    );

    private enum Position {
        LEFT, RIGHT
    }

    @NonNull
    private final Position position;

    public NoiseCancellationListController(@NonNull Context context,
                                           @NonNull String preferenceKey) {
        super(context, preferenceKey);

        position = Position.valueOf(preferenceKey.split("_")[0].toUpperCase());
    }

    @Override
    public int getConfigId() {
        return EarbudsConstants.XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_LIST;
    }

    @Override
    public int getExpectedConfigLength() {
        return 2;
    }

    @NonNull
    @Override
    protected List<ConfigState> getConfigStates() {
        return CONFIG_STATES;
    }

    @NonNull
    @Override
    protected List<ConfigState> getCheckedStates() {
        if (DEBUG) Log.d(TAG, "getCheckedStates");

        if (getConfigValue() == null || !isValidValue(getConfigValue())) {
            return List.of();
        }

        byte configByte = getConfigValue()[position == Position.LEFT ? 0 : 1];

        return CONFIG_STATES.stream()
                .filter(state -> (configByte & (1 << state.configValue[0])) > 0)
                .toList();
    }

    @Override
    public boolean onPreferenceChange(@NonNull Preference preference, Object o) {
        if (DEBUG) Log.d(TAG, "onPreferenceChange: " + o);

        return ((Set<?>) o).size() >= 2;
    }

    @Override
    public boolean saveConfig(@NonNull MMADevice device, @NonNull Object value) throws IOException {
        if (DEBUG) Log.d(TAG, "saveConfig: " + value);

        byte valueByte = ((Set<?>) value).stream()
                .map(s -> 1 << Integer.parseInt((String) s))
                .reduce(0, Integer::sum).byteValue();
        value = position == Position.LEFT
                ? new byte[]{valueByte, VALUE_NOT_MODIFIED}
                : new byte[]{VALUE_NOT_MODIFIED, valueByte};

        return super.saveConfig(device, value);
    }

}

File: src/org/lineageos/xiaomi_bluetooth/configs/NoiseCancellationListController.kt
package org.lineageos.xiaomi_bluetooth.configs

import android.content.Context
import android.util.Log
import androidx.preference.Preference
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_LIST
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_OFF
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_ON
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_TRANSPARENCY
import org.lineageos.xiaomi_bluetooth.R
import org.lineageos.xiaomi_bluetooth.mma.MMADevice
import java.util.Locale

class NoiseCancellationListController(
    context: Context,
    preferenceKey: String
) : MultiListController(context, preferenceKey), Preference.OnPreferenceChangeListener {

    private enum class Position { LEFT, RIGHT }

    private val position =
        Position.valueOf(
            preferenceKey.split("_".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()[0].uppercase(
                Locale.getDefault()
            )
        )

    override val configId = XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_LIST
    override val expectedConfigLength = 2

    override val configStates = setOf(
        ConfigState(
            byteArrayOf(XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_OFF),
            R.string.noise_cancellation_mode_off
        ),
        ConfigState(
            byteArrayOf(XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_ON),
            R.string.noise_cancellation_mode_on
        ),
        ConfigState(
            byteArrayOf(XIAOMI_MMA_CONFIG_NOISE_CANCELLATION_MODE_TRANSPARENCY),
            R.string.noise_cancellation_mode_transparency
        )
    )

    override val checkedStates: Set<ConfigState>
        get() {
            if (DEBUG) Log.d(TAG, "getCheckedStates")

            if (configValue == null || !isValidValue(configValue)) {
                return setOf()
            }

            val configByte = configValue!![if (position == Position.LEFT) 0 else 1]

            return configStates.filter { state ->
                (configByte.toInt() and (1 shl state.configValue[0].toInt())) > 0
            }.toSet()
        }

    override fun onPreferenceChange(preference: Preference, o: Any): Boolean {
        if (DEBUG) Log.d(TAG, "onPreferenceChange: $o")

        return (o as Set<*>).size >= 2
    }

    override fun saveConfig(device: MMADevice, value: Any): Boolean {
        if (DEBUG) Log.d(TAG, "saveConfig: $value")

        val valueByte = (value as Set<*>)
            .map { s -> 1 shl (s as String).toInt() }
            .reduce { a, b -> a + b }
            .toByte()

        val configValue = if (position == Position.LEFT) {
            byteArrayOf(valueByte, VALUE_NOT_MODIFIED)
        } else {
            byteArrayOf(VALUE_NOT_MODIFIED, valueByte)
        }

        return super.saveConfig(device, configValue)
    }

    companion object {
        private val TAG = ListController::class.java.simpleName
        private const val DEBUG = true

        private const val VALUE_NOT_MODIFIED: Byte = -1
    }
}
